fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -134,124 +179,241 @@ impl<T> Interpreter<PredicateStorage, T> {
-                    }
-                    PredicateRunKind::Estimating(tx) => {
-                        let context = Context::PredicateEstimation { program: predicate };
-                        let tx_available_gas = params
-                            .max_gas_per_tx
-                            .checked_sub(cumulative_gas_used)
-                            .ok_or_else(|| Bug::new(BugId::ID003, GlobalGasUnderflow))?;
-                        let available_gas = core::cmp::min(
-                            params.max_gas_per_predicate,
-                            tx_available_gas,
-                        );
-
-                        vm.init_predicate(context, *tx, balances.clone(), available_gas)?;
-                        available_gas
-                    }
-                };
-
-                let result = vm.verify_predicate();
-                let is_successful = matches!(result, Ok(ProgramState::Return(0x01)));
-
-                let gas_used = available_gas
-                    .checked_sub(vm.remaining_gas())
-                    .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?;
-                cumulative_gas_used = cumulative_gas_used
-                    .checked_add(gas_used)
-                    .ok_or_else(|| PredicateVerificationFailed::OutOfGas)?;
-
-                match &mut kind {
-                    PredicateRunKind::Verifying(_) => {
-                        if !is_successful {
-                            result?;
-                            return Err(PredicateVerificationFailed::False)
-                        }
+                vm.init_predicate(context, tx, balances, available_gas)?;
+                available_gas
+            }
+        };
+
+        let result = vm.verify_predicate();
+        let is_successful = matches!(result, Ok(ProgramState::Return(0x01)));
+
+        let gas_used = available_gas
+            .checked_sub(vm.remaining_gas())
+            .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?;
+
+        if let PredicateAction::Verifying = predicate_action {
+            if !is_successful {
+                result?;
+                return Err(PredicateVerificationFailed::False)
+            }
+
+            if vm.remaining_gas() != 0 {
+                return Err(PredicateVerificationFailed::GasMismatch)
+            }
+        }
 
-                        if vm.remaining_gas() != 0 {
-                            return Err(PredicateVerificationFailed::GasMismatch)
+        Ok(gas_used)
+    }
+
+    fn finalize_check_predicate<Tx>(
+        mut kind: PredicateRunKind<Tx>,
+        checks: Vec<Result<Word, PredicateVerificationFailed>>,
+        predicate_action: PredicateAction,
+        params: ConsensusParameters,
+    ) -> Result<PredicatesChecked, PredicateVerificationFailed>
+    where
+        Tx: ExecutableTransaction,
+    {
+        if let PredicateRunKind::Estimating(tx) = &mut kind {
+            checks.iter().enumerate().for_each(|(index, gas_used)| {
```

Comment:
If some inputs are not predicates, the `index` will mismatch, and you will update/skip some `gas_used`.

Also, it would be nice if you add a new unit test to cover this case(it also will validate the fix).

Reply:
good find!, I opted for returning `index` as well from the parallel task that way we are 100% it's the index of the correct inputs array